beacon-chain/sync/data_column_sidecars.go

```diff
@@ -39,249 +39,394 @@ type DataColumnSidecarsParams struct {
+	indirectSidecarsByRoot, err := requestIndirectSidecarsFromPeers(params, roBlocks, slotsWithCommitments, storedIndicesByRoot, directSidecarsByRoot, requestedIndices, incompleteRoots)
 	if err != nil {
-		return nil, errors.Wrap(err, "request from peers")
+		return nil, nil, errors.Wrap(err, "request all sidecars from peers")
 	}
 
-	log.WithFields(logrus.Fields{"duration": time.Since(start), "count": count}).Debug("Requested data column sidecars from peers")
+	// Merge sidecars in storage and those received from peers. Reconstruct if needed.
+	mergedSidecarsByRoot, err = mergeStorageAndInputs(params.Storage, storedIndicesByRoot, indirectSidecarsByRoot, requestedIndices, incompleteRoots)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "try merge storage and all inputs")
+	}
 
-	// Step 3c: If needed, try to reconstruct missing sidecars from storage and fetched data.
-	fromReconstructionResult, err := tryReconstructFromStorageAndPeers(params.Storage, fromPeersResult, indicesMap, missingIndicesByRoot)
+	for root, sidecars := range mergedSidecarsByRoot {
+		result[root] = sidecars
+	}
+
+	if len(incompleteRoots) == 0 {
+		log.WithField("finalMissingRootCount", len(incompleteRoots)).Debug("Fetched data column sidecars from storage and peers using rescue mode")
+		return result, nil, nil
+	}
+
+	// For remaining incomplete roots, assemble what is available.
+	incompleteSidecarsByRoot, missingByRoot, err := assembleAvailableSidecars(incompleteRoots, params.Storage, requestedIndices, directSidecarsByRoot)
 	if err != nil {
-		return nil, errors.Wrap(err, "reconstruct from storage and peers")
+		return nil, nil, errors.Wrap(err, "assemble available sidecars for incomplete roots")
 	}
 
-	for root, verifiedSidecars := range fromReconstructionResult {
-		result[root] = verifiedSidecars
+	for root, sidecars := range incompleteSidecarsByRoot {
+		result[root] = sidecars
 	}
 
-	for root := range fromPeersResult {
-		if _, ok := fromReconstructionResult[root]; ok {
-			// We already have what we need from peers + reconstruction
+	log.WithField("finalMissingRootCount", len(incompleteRoots)).Debug("Failed to fetch data column sidecars from storage and peers using rescue mode")
+	return result, missingByRoot, nil
+}
+
+// commitmentsInfo returns:
+// - a set of slots for blocks with commitments,
+// - a map root->stored indices for blocks with commitments, and
+// - a set of roots for blocks with commitments.
+func commitmentsInfo(storage filesystem.DataColumnStorageReader, roBlocks []blocks.ROBlock) (map[primitives.Slot]bool, map[[fieldparams.RootLength]byte]map[uint64]bool, map[[fieldparams.RootLength]byte]bool, error) {
+	blockCount := len(roBlocks)
+
+	slotsWithCommitments := make(map[primitives.Slot]bool, blockCount)
+	storedIndicesByRoot := make(map[[fieldparams.RootLength]byte]map[uint64]bool, blockCount)
+	roots := make(map[[fieldparams.RootLength]byte]bool, blockCount)
```

Comment:
why is this called an incomplete root?

Reply:
Fixed in https://github.com/OffchainLabs/prysm/commit/0f0bcf1237e61ee04d4b1ca1e91c23ec2345608d.